### PR TITLE
Fix config overwriting

### LINKED ISSUE #879: Covalent config is improperly overwritten
### Environment

- **Covalent version**: 0.142.0
- **Python version**: 3.8
- **Operating system**: Gentoo

### What is happening?

The dispatcher address is improperly overwritten when the server is started.  I'm not sure if it also happens to other values in the config.

### How can we reproduce the issue?

1. Generate a default config file, e.g., using `covalent start`.
2. Stop the Covalent server with `covalent stop`
3. Manually edit the address in the config.
4. Start the server with `covalent start`.
5. Inspect the config to see "localhost" has overwritten the provided value.

### What should happen?

A user's config file should take precedence over any default values.  In addition, the server always seems to start on localhost, which is hard-coded into `covalent_ui/app.py`, even though it should really reference the config file.  E.g. in my case I want it to start on `0.0.0.0` instead of `localhost`.

### Any suggestions?

_No response_

### PR DESCRIPTION
**Pull request by FyzHsn**
Fixes #879 

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation and CHANGELOG accordingly.
- [x] I have read the CONTRIBUTING document.

### CODE REVIEW

**Review by venkatBala — DISMISSED** [review]

File: covalent_dispatcher/_cli/service.py [hunk 1]
```diff
@@ -320,22 +320,31 @@ def start(
...
-            "dispatcher.address": "localhost",
             "dispatcher.port": port,
-            "dask": {
-                "mem_per_worker": mem_per_worker or "auto",
-                "threads_per_worker": threads_per_worker or 1,
-                "num_workers": workers or dask.system.CPU_COUNT,
-            },
         }
     )
+    if no_cluster_flag not in sys.argv:
```
Comment (wjcunningham7) [inline covalent_dispatcher/_cli/service.py:330]: I think the logic we want is this:
```
if not no_cluster:
    if mem_per_worker != -1:
        set_config({"dask": {"mem_per_worker": mem_per_worker}})
    if threads_per_worker != -1:
        set_config({"dask": {"threads_per_worker": threads_per_worker}})
    if workers != -1:
        set_config({"dask": {"num_workers": workers}})
```

and then for each of those three variables, the click defaults should be set to -1.  The actual defaults that you want in the config should then be modified in the `dask` block inside `covalent/_shared_files/defaults.py`.

Comment (FyzHsn) [reply covalent_dispatcher/_cli/service.py]: @wjcunningham7 

I checked the `if no_cluster_flag not in sys.argv:` vs `if not no_cluster:` and they both work.

I just added the updated logic according to your comment. Also note that there's some nuance when the defaults haven't been set yet in the config and the user may or may not have specified some default. I'll probably do some refactoring to reduce some of the code bloat here if you approve the logic

File: covalent_dispatcher/_cli/service.py [hunk 2]
```diff
@@ -320,22 +320,70 @@ def start(
...
-            "dask": {
-                "mem_per_worker": mem_per_worker or "auto",
-                "threads_per_worker": threads_per_worker or 1,
-                "num_workers": workers or dask.system.CPU_COUNT,
-            },
         }
     )
 
+    if not no_cluster:
+        if not mem_per_worker:
```
Comment (wjcunningham7) [inline covalent_dispatcher/_cli/service.py:331]: I think this should be `if mem_per_worker`

File: covalent_dispatcher/_cli/service.py [hunk 3]
```diff
@@ -253,7 +253,7 @@ def _graceful_shutdown(pidfile: str) -> None:
     required=False,
     is_flag=False,
     type=str,
-    default="auto",
+    # default="auto",
```
Comment (wjcunningham7) [inline covalent_dispatcher/_cli/service.py:256]: are the defaults "" for strings and 0 for ints?  If so you can remove these commented lines

**Review by wjcunningham7 — CHANGES_REQUESTED** [review]

File: covalent_dispatcher/_cli/service.py (hunk 3 above)
Comment (venkatBala) [reply covalent_dispatcher/_cli/service.py:256]: the default needed for `mem_per_worker` is "auto" and `threads_per_worker =1`, and `num_workers=dask.system.CPU_COUNT`

File: covalent_dispatcher/_cli/service.py [hunk 4]
```diff
@@ -320,22 +320,70 @@ def start(
...
-                "mem_per_worker": mem_per_worker or "auto",
-                "threads_per_worker": threads_per_worker or 1,
-                "num_workers": workers or dask.system.CPU_COUNT,
-            },
         }
     )
 
+    if not no_cluster:
+        if not mem_per_worker:
+            set_config(
```
Comment (wjcunningham7) [inline covalent_dispatcher/_cli/service.py:332]: proper way to call this is `set_config("dask.mem_per_worker", mem_per_worker)`

**Comment by venkatBala** [comment]
@wjcunningham7 I've made the necessary changes as per discussion this morning.

@FyzHsn Kindly review the logic and see if everything is upto spec

**Comment by FyzHsn** [comment]
> @wjcunningham7 I've made the necessary changes as per discussion this morning.
> 
> @FyzHsn Kindly review the logic and see if everything is upto spec

Yessir, the logic is looking good! I'll update the tests to meet the diff coverage requirement.

**Review by wjcunningham7 — DISMISSED** [review]

**Review by wjcunningham7 — DISMISSED** [review]

**Review by venkatBala — DISMISSED** [review]

**Review by wjcunningham7 — APPROVED** [review]

**Pull request merged**